node/src/components/consensus/highway_core/state/index_panorama.rs

```diff
@@ -0,0 +1,56 @@
+use std::fmt::Debug;
+
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+use crate::components::consensus::{
+    highway_core::{
+        state::{Observation, Panorama, State},
+        validators::ValidatorMap,
+    },
+    traits::Context,
+};
+
+pub(crate) type IndexPanorama = ValidatorMap<IndexObservation>;
+
+/// The observed behavior of a validator at some point in time.
+#[derive(Clone, DataSize, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub(crate) enum IndexObservation {
+    None,
+    Faulty,
+    Correct(u64),
+}
+
+impl Debug for IndexObservation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            IndexObservation::None => write!(f, "N"),
+            IndexObservation::Faulty => write!(f, "F"),
+            IndexObservation::Correct(seq_num) => write!(f, "{:?}", seq_num),
+        }
+    }
+}
+
+impl IndexPanorama {
+    /// Creates an instance of `IndexPanorama` out of a panorama.
+    pub(crate) fn from_panorama<'a, C: Context>(
+        panorama: &'a Panorama<C>,
+        state: &'a State<C>,
+    ) -> Self {
+        let mut validator_map: ValidatorMap<IndexObservation> =
+            ValidatorMap::from(vec![IndexObservation::None; panorama.len()]);
+        for (vid, obs) in panorama.enumerate() {
+            let index_obs = match obs {
+                Observation::None => IndexObservation::None,
+                Observation::Correct(hash) => state
+                    .maybe_unit(hash)
+                    .map_or(IndexObservation::None, |unit| {
```

Comment:
I'm not sure about this part of the code - it seems like it should have been a fatal error. It's a case where our panorama claims we should have that unit in the state but then `state` doesn't have it. I'm on the fence here about panic-ing and crashing the node. I chose to _not_ do that as the `IndexPanorama::from_panorama` method is used only in the "secondary" method of protocol synchronization which didn't seem "important" enough to justify shutting down the whole operation. \

Having written that - I will at least add an `error`-level logging here.

Reply:
I was thinking of commenting on that, but this function can take an arbitrary panorama, so us not having the indicated unit isn't necessarily a fatal error.
It might be a good idea to return an error in that case, and then panic at the call site if it was our own panorama that was passed to the function.

Reply:
>I was thinking of commenting on that, but this function can take an arbitrary panorama, so us not having the indicated unit isn't necessarily a fatal error.

That is a good point. Initially, the panorama was taken from the `state` that we're passing in but I've changed it so that it accepts an arbitrary panorama.